packages/react-ui/src/TableOfContent/useTOCReducer.js

```diff
@@ -0,0 +1,32 @@
+import { useReducer, useCallback } from "react";
+
+export const rootReducer = (state, action) => {
+  switch (action.type) {
+    case "setTitles":
+      return { titles: [...action.payload.map((title) => ({ ...title }))] };
+    case "setActive":
+      return {
+        titles: [
+          ...state.titles.map((title) => ({
+            ...title,
+            active: title.id === action.payload ? true : false,
+          })),
+        ],
+      };
+  }
+};
+
+export const useTOCReducer = (initialState) => {
+  const [state, dispatch] = useReducer(rootReducer, initialState);
+  const observerCallback = useCallback((entries) => {
+    entries.forEach((entry) => {
+      if (entry.intersectionRatio === 1) {
+        dispatch({ type: "setActive", payload: entry.target.id });
+      }
+    });
+  }, []);
+  const setTitles = useCallback((titles) => {
```

Comment:
I don't understand why do you wrap the function with a useCallback.

Reply:
Alors j'ai peur de dire une connerie mais en fait à la base le reducer était pas dans un fichier à part et observerCallback était défini dans le composant TOC. A chaque re-render  la fonction était donc redéfinie et je l'ai mise dans un useCallback.
Là c'est un peu caché parce que c'est dans un fichier à part mais il faut se dire que "useTOCReducer" utilise des hooks certes, mais reste appelé à chaque re-render. Ce sont les hooks dedans qui ne sont pas réappelés à chaque fois. Donc le useCallback permet de ne pas redéfinir la fonction à chaque fois. Après est-ce que ça fonctionne là ? je pense mais il faudrait que je teste pour être sur sauf que c'est pas facile de tester dans le storybook, il faudrait que je teste à l'aide de la PR d'après 🤷